arelle/XbrlUtil.py

```diff
@@ -135,45 +168,60 @@ def attributeDict(modelXbrl, elt, exclusions=set(), equalMode=S_EQUAL, excludeID
                 pass  # what should be done if attribute failed to have psvi value
     return attrs
 
-def attributes(modelXbrl, elt, exclusions=set(), ns2ns1Tbl=None, keyByTag=False) -> tuple[Any, Any]:
+def attributes(
+    modelXbrl: ModelXbrl,
+    elt: ModelObject,
+    exclusions: set[str] = set(),
+    ns2ns1Tbl: dict[str, str] | None = None,
+    keyByTag: bool = False
+) -> tuple[tuple[QName, Any], ...]:
     a = attributeDict(modelXbrl, elt, exclusions, ns2ns1Tbl=ns2ns1Tbl, keyByTag=keyByTag)
     return tuple( (k,a[k]) for k in sorted(a.keys()) )
 
-def childElements(elt):
+def childElements(elt: ModelObject) -> list[ModelObject]:
     return [child for child in elt if isinstance(child,ModelObject)]
 
-def xEqual(elt1, elt2, equalMode=S_EQUAL):
+def xEqual(elt1: ModelObject, elt2: ModelObject, equalMode: int = S_EQUAL) -> bool:
     if not hasattr(elt1,"xValid"):
-        xmlValidate(elt1.modelXbrl, elt1)
+        xmlValidate(elt1.modelXbrl, elt1)  # type: ignore[no-untyped-call]
     if not hasattr(elt2,"xValid"):
-        xmlValidate(elt2.modelXbrl, elt2)
+        xmlValidate(elt2.modelXbrl, elt2)  # type: ignore[no-untyped-call]
     if equalMode == VALIDATE_BY_STRING_VALUE:
         return elt1.stringValue == elt2.stringValue
     elif equalMode == S_EQUAL: # formula WG e-mail 2018-09-06: or (equalMode == S_EQUAL2 and not isinstance(elt1.sValue, QName)):
         return elt1.sValue == elt2.sValue
     else: # includes dimension S-equal2, use xpath-2 equality.
-        if isinstance(elt1.xValue, DateTime) and isinstance(elt2.xValue, DateTime) and elt1.xValue.dateOnly != elt2.xValue.dateOnly:
+        elt1_xValue = elt1.xValue
+        elt2_xValue = elt2.xValue
+        if isinstance(elt1_xValue, DateTime) \
+            and isinstance(elt2_xValue, DateTime) \
+                and getattr(elt1_xValue, 'dateOnly') \
```

Comment:
This should be non-nullable even if its not now, its just a bool right?

Reply:
`dateOnly` gets assigned with __new__ of `DateTime`, mypy doesn't pick that up, maybe its better to add it as class attribute in #329, I will ingore[attr-defined] for now here.